testdata/testoutput2

```diff
@@ -20646,10 +20765,151 @@ Failed: error 150 at offset 5: invalid range in character class
 No match
 
 /[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[z]&&a]]]]]]]]]]]]]]]/alt_extended_class
-Failed: error 207 at offset 115: character classes are too deeply nested
+Failed: error 207 at offset 115: character class nesting is too deep
 
 /[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a[b]&&a[a&&a[z]]]]]]]]]]]]]]]]/alt_extended_class
-Failed: error 207 at offset 118: character classes are too deeply nested
+Failed: error 207 at offset 118: character class nesting is too deep
+
+# --------------
+
+# EXTENDED CHARACTER CLASSES (Perl)
+
+# allow-empty-class does nothing inside (?[...])
+/(?[ []] ])/allow_empty_class
+    ]
+ 0: ]
+
+# bad-escape-is-literal does nothing inside (?[...])
+/[ \j ]/
+Failed: error 103 at offset 3: unrecognized character follows \
+
+/(?[ \j ])/
+Failed: error 103 at offset 5: unrecognized character follows \
+
+/[ \j ]/bad_escape_is_literal
+    j
+ 0: j
+\= Expect no match
+    k
+No match
+
+/(?[ \j ])/bad_escape_is_literal
+Failed: error 103 at offset 5: unrecognized character follows \
+
+/(?[ [\j] ])/bad_escape_is_literal
+Failed: error 103 at offset 6: unrecognized character follows \
+
+/(?[ (\j) ])/bad_escape_is_literal
+Failed: error 103 at offset 6: unrecognized character follows \
+
+# We can't test error cases in testinput1
+
+/(?[])/
+Failed: error 214 at offset 4: empty expression in character class
+
+/(?[/
+Failed: error 106 at offset 3: missing terminating ] for character class
+
+/(?[]/
+Failed: error 214 at offset 4: empty expression in character class
+
+/(?[\n/
+Failed: error 106 at offset 5: missing terminating ] for character class
+
+/(?[\n]/
+Failed: error 215 at offset 6: unexpected ] with no following parenthesis in (?[...]
+
+/(?[\n]z)/
+Failed: error 215 at offset 6: unexpected ] with no following parenthesis in (?[...]
+
+/(?[\n] )/
+Failed: error 215 at offset 6: unexpected ] with no following parenthesis in (?[...]
+
+/(?[ \n + () ])/
+Failed: error 214 at offset 11: empty expression in character class
+
+/(?[1])/
+Failed: error 216 at offset 4: unexpected character in (?[...]) character class
+
+/(?[a])/
+Failed: error 216 at offset 4: unexpected character in (?[...]) character class
+
+/(?[a-c])/
+Failed: error 216 at offset 4: unexpected character in (?[...]) character class
+
+/(?[(])/
+Failed: error 114 at offset 4: missing closing parenthesis
```

Comment:
Isn't parenthesis are normal characters in a class?

Reply:
Is it the same as '[' ? Or does it do something different?

Reply:
In a Perl extended class, `(` and `[` are quite different. There are two parsing grammars, the "outer" expression grammar, and an "inner" character class grammar.

```ebnf
PerlExtended := "(?[" OuterExpr "])" ;

OuterExpr := OuterExpr2 ([|+^-] OuterExpr2)* ;
OuterExpr2 := OuterExpr3 ([&] OuterExpr3)* ;
OuterExpr3 := [!]* OuterExpr4 ;
OuterExpr4 := InnerClass | "(" OuterExpr ")" ;

InnerClass := "\\" [dDsSwW] | "[:" Posix ":]" | "[" NormalPerlCharacterClassSyntax "]" ;
```

So `(` doesn't change the parse mode, it stays within the expression syntax (in which "-" is an operator). But "[" switches the parse mode to the ordinary Perl class syntax (in which "-" is a character range, maybe).